[flang] Add basic -mtune support

AlexisPerry opened the pull request:
This PR adds -mtune as a valid flang flag and passes the information through to LLVM IR as an attribute on all functions.  No specific architecture optimizations are added at this time.

AlexisPerry committed:
[flang] Add basic -mtune support

AlexisPerry committed:
clang-format edit

Dinistro commented on mlir/include/mlir/Dialect/LLVMIR/LLVMAttrDefs.td:
```diff
@@ -58,6 +58,15 @@ def FramePointerKindAttr : LLVM_Attr<"FramePointerKind", "framePointerKind"> {
+// TuneCPUAttr
+//===----------------------------------------------------------------------===//
+
+//def TuneCPUAttr : LLVM_Attr<"TuneCPU", "tuneCPU"> {
+  //let parameters = (ins "tuneCPU::tuneCPU":$tuneCPU);
+  //let assemblyFormat = "`<` $tuneCPU `>`";
+//}
```
Dead code?

Dinistro commented on clang/lib/Driver/ToolChains/Flang.cpp:
```diff
@@ -411,6 +412,13 @@ void Flang::addTargetOptions(const ArgList &Args,
   // TODO: Add target specific flags, ABI, mtune option etc.
+  if (const Arg *A = Args.getLastArg(options::OPT_mtune_EQ)) {
+    CmdArgs.push_back("-tune-cpu");
+    if (strcmp(A->getValue(), "native") == 0)
```
Nit: `strcmp` is a C function. I assume that `A->getValue()` returns a StringRef, or something else that supports normal `==` comparison.

AlexisPerry replied:
I see your point, however when I build the code using normal `==` comparison I get a warning:
`warning: comparison with string literal results in unspecified behavior`
and I am inclined to avoid such possible unspecified behavior.

tarunprabhu replied:
`A` is an `llvm::Arg`. The `getValue()` method [returns](https://llvm.org/doxygen/classllvm_1_1opt_1_1Arg.html#ac268590692356db84db78050196b4940) a `const char*`, so unless it is wrapped in a `llvm::StringRef`, the `==` operator cannot be used for string comparison.

It is surprising though. I, too, would have expected it to return a `llvm::StringRef`.

Dinistro replied:
A few lines further up, the string literal get's wrapped into a `StringRef`, which seems to work. 

`if (A->getValue() == StringRef{"Accelerate"})`

AlexisPerry replied:
Thank you for that!  I have made the requested update and there was no build warning.

Dinistro reviewed (commented):
This is missing tests for the LLVM import and export. I suspect that this is currently still part of the function's passthrough dictionary, which should be changed.

tarunprabhu commented on flang/tools/bbc/bbc.cpp:
```diff
@@ -371,7 +371,7 @@ static mlir::LogicalResult convertFortranSourceToMLIR(
       semanticsContext.targetCharacteristics(), parsing.allCooked(),
       targetTriple, kindMap, loweringOptions, envDefaults,
-      semanticsContext.languageFeatures(), targetMachine);
+      semanticsContext.languageFeatures(), targetMachine, ""); // FIXME
```
What needs to be fixed here? If it needs to be deferred to a later patch, a more detailed inline comment would be helpful.

AlexisPerry replied:
Thank you for catching that.  I meant to remove that comment as the thing it was referring to has been fixed.  I'll clean it up now.

tarunprabhu reviewed (commented):
Is there a way to test that the attribute is lowered to FIR and perhaps also on to LLVM? `flang/test/Lower/` has some tests which check for the presence of certain constructs in the LLVM IR. Could something be written similar to those?

banach-space reviewed (commented):
Thanks for working on this @AlexisPerry ! Could you add some tests?

AlexisPerry committed:
Address review comments - code cleanup

AlexisPerry committed:
added mtune tests

AlexisPerry committed:
Merge branch 'main' into AlexisPerry/mtune-2

tarunprabhu commented on flang/test/Lower/tune-cpu-fir.f90:
Why do you have the first `target-cpu` check on its own? Is it intended to check for the _absence_ of the `tune-cpu` attribute when `target-cpu` is provided on the command line? If so, there should be a `CPU-NOT` annotation as well. Otherwise, it is odd to test for `target-cpu` when this file is intended to test `tune-cpu`. 

The third test which checks the case when both `target -cpu` and `tune-cpu` are provided is good.

AlexisPerry replied:
Thank you for the catch! Yes, the CPU-NOT line was missing.  I have rectified this.

tarunprabhu commented on flang/test/Lower/tune-cpu-llvm.f90:
Consider replacing `#0` with `#{{[0-9]+}}`. While the attributes are almost certainly going to be in `#0` for such a small function, you don't actually _require_ this to match `#0`, so you it might be better to let it match any number. 

Also, if the `tune-cpu` attribute is expected to be on the same line, you could use `CHECK-SAME`.

AlexisPerry replied:
Those are great suggestions, thank you.  I have made the appropriate updates.

AlexisPerry committed:
updates to mtune tests following reviewer feedback

tarunprabhu commented on flang/tools/bbc/bbc.cpp:
```diff
@@ -371,7 +371,7 @@ static mlir::LogicalResult convertFortranSourceToMLIR(
       semanticsContext.targetCharacteristics(), parsing.allCooked(),
       targetTriple, kindMap, loweringOptions, envDefaults,
-      semanticsContext.languageFeatures(), targetMachine);
+      semanticsContext.languageFeatures(), targetMachine, "");
```
NIT: Would it be worth creating a `constexpr const char* tuneCPU = ""` and using that in place of the empty string literal here? It might make it clear what that parameter is. 

I don't have a strong opinion on this one way or another. If you or anyone else has a different opinion, I'm happy to defer to it.

AlexisPerry replied:
Thank you Tarun, I think that's a much better way to do this. I have made the change.

gysit commented on mlir/lib/Target/LLVMIR/ModuleImport.cpp:
```diff
@@ -1796,6 +1797,10 @@ void ModuleImport::processFunctionAttributes(llvm::Function *func,
 
+  if (llvm::Attribute attr = func->getFnAttribute("tune-cpu");
+      attr.isStringAttribute())
+    funcOp.setTuneCpuAttr(StringAttr::get(context, attr.getValueAsString()));
```
Could you also add a small test that verifies the import / export works independent from Flang?

The target_cpu attributes are tested here:
```
llvm-project/mlir/test/Target/LLVMIR/target-cpu.mlir
llvm-project/mlir/test/Target/LLVMIR/Import/target-cpu.ll
```
It may make sense to add the tests in the same files and maybe rename them to target-and-tune-cpu.mlir/ll or have a separate set of files if the attributes are unrelated (which is not my impression).

Thanks!

AlexisPerry replied:
Thank you so much for this suggestion and for the pointers for how to do this kind of test.  I was struggling to figure out the proper mechanism and you helped greatly.  I have added two separate tests for tune-cpu instead of combining with the target-cpu ones because I think it's cleaner, but I am happy to combine if you prefer. Just let me know.  Thanks again for your help!

gysit replied:
Nice!

Separate tests are fine.

AlexisPerry committed:
Address review comments - variable name change

AlexisPerry commented on flang/include/flang/Frontend/TargetOptions.h:
```diff
@@ -32,6 +32,9 @@ class TargetOptions {
   std::string cpu;
 
+  /// If given, the name of the target CPU to tune code for.
+  std::string tuneCPU;
```
Thank you for the suggestion.  I have made the change. :-)

AlexisPerry committed:
added mlir import/export tests for tune-cpu

AlexisPerry committed:
Address review comments - increase clarity in bbc default lowering

AlexisPerry committed:
clang-format

tarunprabhu reviewed (commented):
Thanks for the changes. LGTM, but wait for @banach-space to approve.

Dinistro commented on mlir/test/Target/LLVMIR/Import/tune-cpu.ll:
```diff
@@ -0,0 +1,9 @@
+; RUN: mlir-translate -import-llvm -split-input-file %s | FileCheck %s
+
+; CHECK-LABEL: llvm.func @tune_cpu()
+; CHECK-SAME: tune_cpu = "pentium4"
```
Can you add a check that ensures that this is not also placed in the passthrough dictionary? I don't see any part in the code that would avoid this. Adding it to the `ExplicitAttributes` set in `ModuleImport.cpp` should be enough.

AlexisPerry replied:
From what I can see, L1685 of ModuleImport.cpp has tune-cpu listed among the ExplicitAttributes.  Is this sufficient?  Or is further checking needed?  Thank you so much.

Dinistro replied:
Given that we are both not entirely sure what happens, it might be sensible to check for that, just to be sure 🙂

AlexisPerry replied:
Alright, after some digging, I can confirm that `tune-cpu` is listed among the `kExplicitAttributes` and that these attributes are expressly excluded from being added to the passthrough dictionary based on the code in `processPassthroughAttrs` in `ModuleImport.cpp`.

banach-space commented on flang/test/Lower/tune-cpu-fir.f90:
This is probably a border-line case, but I would consider this a "driver" rather than a "lowering" test. I'm biased though 😅

AlexisPerry replied:
I hear you.  I modeled this test closely after target-features-*.f90 in that same directory, which is why I put it in flang/test/Lower.  I am willing to move it to flang/test/Driver, but if I do so should I also move the target-features-*.f90 tests?  Thank you so much.

banach-space replied:
> if I do so should I also move the target-features-*.f90 tests?

Yes, but to me that would qualify as an "unrelated change" (i.e. sth for a separate PR, no need to worry about it here). 

In general, this PR is about enabling a flag in Flang and making sure that it actually works. This kind of "integration" is a driver task, hence my suggestion.

AlexisPerry replied:
Thanks for explaining more thoroughly.  I have moved the test to the Driver directory as requested.

banach-space reviewed (approved):
LGTM, thanks for implementing this 🙏🏻

AlexisPerry committed:
Address reviewer comments - remove strcmp usage

AlexisPerry commented:
I have lost my upstream commit privileges due to inactivity, so once this is fully approved, could someone merge it on my behalf?  Thank you.

Dinistro reviewed (approved):
The MLIR side looks good to me % one remaining nit comment. Thanks for addressing the comments.

AlexisPerry committed:
Address review comments - move test from Lower to Driver

AlexisPerry commented:
I believe I've addressed all the review comments and all the checks have passed.  Could someone with commit access please merge this on my behalf?  Thank you.

banach-space commented:
> LLVM Buildbot has detected a new failure on builder `flang-aarch64-libcxx` running on `linaro-flang-aarch64-libcxx` while building `clang,flang,mlir` at step 6 "test-build-unified-tree-check-flang".
> 
> Full details are available at: https://lab.llvm.org/buildbot/#/builders/89/builds/791
> 
> Here is the relevant piece of the build log for the reference:
> 
> ```
> Step 6 (test-build-unified-tree-check-flang) failure: test (failure)
> ******************** TEST 'Flang :: Lower/tune-cpu-llvm.f90' FAILED ********************
> Exit Code: 2
> 
> Command Output (stderr):
> --
> RUN: at line 1: /home/tcwg-buildbot/worker/flang-aarch64-libcxx/build/bin/flang-new -mtune=pentium4 -S -emit-llvm /home/tcwg-buildbot/worker/flang-aarch64-libcxx/llvm-project/flang/test/Lower/tune-cpu-llvm.f90 -o - | /home/tcwg-buildbot/worker/flang-aarch64-libcxx/build/bin/FileCheck /home/tcwg-buildbot/worker/flang-aarch64-libcxx/llvm-project/flang/test/Lower/tune-cpu-llvm.f90
> + /home/tcwg-buildbot/worker/flang-aarch64-libcxx/build/bin/flang-new -mtune=pentium4 -S -emit-llvm /home/tcwg-buildbot/worker/flang-aarch64-libcxx/llvm-project/flang/test/Lower/tune-cpu-llvm.f90 -o -
> + /home/tcwg-buildbot/worker/flang-aarch64-libcxx/build/bin/FileCheck /home/tcwg-buildbot/worker/flang-aarch64-libcxx/llvm-project/flang/test/Lower/tune-cpu-llvm.f90
> flang-new: error: unsupported argument 'pentium4' to option '-mtune='
> FileCheck error: '<stdin>' is empty.
> FileCheck command line:  /home/tcwg-buildbot/worker/flang-aarch64-libcxx/build/bin/FileCheck /home/tcwg-buildbot/worker/flang-aarch64-libcxx/llvm-project/flang/test/Lower/tune-cpu-llvm.f90
> 
> --
> 
> ********************
> ```

Most likely https://github.com/llvm/llvm-project/pull/95043/files#diff-a29a79ef4763ed66987d979a7a8a4ff87d242101fe133d5188577b9ff144b805 requires X86 to be enabled. Could somebody either fix or revert? I don’t have access to Git ATM.
